### DISCUSSION TITLE
A general library for higher-kinded data

### DISCUSSION THREAD

**Discussion by chris-martin** [circuithub/rel8#124] [Ideas]
Rel8 defines the term "higher-kinded data" as:

> Higher-kinded data types are data types of the pattern:
>
> ```haskell
> data MyType f =
>   MyType { field1 :: Column f T1 OR HK1 f
>          , field2 :: Column f T2 OR HK2 f
>          , ...
>          , fieldN :: Column f Tn OR HKn f
>          }
> ```
> 
> where Tn is any Haskell type, and HKn is any higher-kinded type.

Aside from the specific detail of the `Column` type family, this notion seems very general! Do y'all think there's any potential for a library that would help others replicate the pattern for other applications? (For example, I have in mind a similar sort of approach to filesystem storage, with database rows and fields corresponding to directories and files.)

**Comment by ocharles** [comment]
Maybe! Though I will note I think those libraries maybe already exist. In this space, we have:

1. https://hackage.haskell.org/package/barbies - probably the one I'd most recommend
2. https://hackage.haskell.org/package/rank2classes
3. https://hackage.haskell.org/package/hkd

and there may be others. So I think we'd have to be confident that we have something substantially different. There's a good chance that if https://ghc-proposals.readthedocs.io/en/latest/proposals/0242-unsaturated-type-families.html makes it into GHC proper, we can drop a lot of what we've had to invent in Rel8 and just use `barbies` - but that remains to be proven.

**Comment by ocharles** [comment]
(@chris-martin I moved this to a discussion - hope that's OK!)

**Reply by chris-martin** [reply]
> (@chris-martin I moved this to a discussion - hope that's OK!)

Oh cool, I didn't even know about github discussions. Yes, it's certainly not a rel8 "issue".

**Comment by chris-martin** [comment]
The main things I really like about `Rel8able` that I haven't managed to replicate more generally are:

  - The `Column` type family enumerates the library-specific type constructors that inhabit `f`, which makes the API much easier to understand than directly dealing with general-purpose type variable soups like `(forall a. p a -> m (q a)) -> g p -> m (g q)`
  - The type family approach allows the `Identity` type to disappear
  - You derive the library-specific `Rel8able` class, rather than something generic like `Rank2.Traversable`, which means if there are any mistakes in the data definition, you see the type errors immediately where the datatype is defined, rather than seeing some constraint violation off at some use site.

By the way, I'd found `barbies` and `hkd`, but thanks for pointing out `rank2classes` - that one is working out much better for me.
